x/gasless/keeper/fee_helper.go

```diff
@@ -0,0 +1,241 @@
+package keeper
+
+import (
+	"strconv"
+	"strings"
+
+	sdkerrors "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension-rdk/x/gasless/types"
+)
+
+func (k Keeper) EmitFeeConsumptionEvent(
+	ctx sdk.Context,
+	feeSource sdk.AccAddress,
+	failedGasTankIDs []uint64,
+	failedGasTankErrors []error,
+	succeededGtid uint64,
+) {
+	failedGasTankIDsStr := make([]string, 0, len(failedGasTankIDs))
+	for _, id := range failedGasTankIDs {
+		failedGasTankIDsStr = append(failedGasTankIDsStr, strconv.FormatUint(id, 10))
+	}
+	failedGasTankErrorMessages := make([]string, 0, len(failedGasTankErrors))
+	for _, err := range failedGasTankErrors {
+		failedGasTankErrorMessages = append(failedGasTankErrorMessages, err.Error())
+	}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeFeeConsumption,
+			sdk.NewAttribute(types.AttributeKeyFeeSource, feeSource.String()),
+			sdk.NewAttribute(types.AttributeKeyFailedGasTankIDs, strings.Join(failedGasTankIDsStr, ",")),
+			sdk.NewAttribute(types.AttributeKeyFailedGasTankErrors, strings.Join(failedGasTankErrorMessages, ",")),
+			sdk.NewAttribute(types.AttributeKeySucceededGtid, strconv.FormatUint(succeededGtid, 10)),
+		),
+	})
+}
+
+func (k Keeper) GetUpdatedGasConsumerAndConsumptionIndex(ctx sdk.Context, consumerAddress sdk.AccAddress, gasTank types.GasTank, feeAmount sdkmath.Int) (types.GasConsumer, uint64) {
+	gasConsumer, consumptionIndex := k.GetOrCreateGasConsumer(ctx, consumerAddress, gasTank)
+	gasConsumer.Consumptions[consumptionIndex].TotalTxsMade = gasConsumer.Consumptions[consumptionIndex].TotalTxsMade + 1
+	gasConsumer.Consumptions[consumptionIndex].TotalFeesConsumed = gasConsumer.Consumptions[consumptionIndex].TotalFeesConsumed.Add(feeAmount)
+	return gasConsumer, consumptionIndex
+}
+
+func (k Keeper) CanGasTankBeUsedAsSource(ctx sdk.Context, gtid uint64, consumer types.GasConsumer, fee sdk.Coin) (gasTank types.GasTank, isValid bool, err error) {
+	gasTank, found := k.GetGasTank(ctx, gtid)
+	// there is no gas tank with given id, likely impossible to happen
+	// exists only as aditional check.
+	if !found {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "gas tank not found")
+	}
```

Comment:
It's either impossible or it's not.
We don't want code that noone understands if it's needed or not.
Needs to be justified

Reply:
fixed